lib/web_ui/dev/steps/run_tests_step.dart

```diff
@@ -0,0 +1,295 @@
+  });
+
+  final BrowserEnvironment browserEnvironment;
+  final List<FilePath>? testFiles;
+  final bool isDebug;
+  final bool doUpdateScreenshotGoldens;
+
+  /// Global list of shards that failed.
+  ///
+  /// This is used to make sure that when there's a test failure anywhere we
+  /// exit with a non-zero exit code.
+  ///
+  /// Shards must never be removed from this list, only added.
+  List<String> failedShards = <String>[];
+
+  /// Whether all test shards succeeded.
+  bool get allShardsPassed => failedShards.isEmpty;
+
+  @override
+  String get description => 'run_tests';
+
+  @override
+  bool get isSafeToInterrupt => true;
+
+  @override
+  Future<void> interrupt() async {}
+
+  @override
+  Future<void> run() async {
+    _copyTestFontsIntoWebUi();
+    await _prepareTestResultsDirectory();
+    await browserEnvironment.prepareEnvironment();
+
+    final List<FilePath> testFiles = this.testFiles ?? findAllTests();
+
+    // Separate screenshot tests from unit-tests. Screenshot tests must run
+    // one at a time. Otherwise, they will end up screenshotting each other.
+    // This is not an issue for unit-tests.
+    final FilePath failureSmokeTestPath = FilePath.fromWebUi(
+      'test/golden_tests/golden_failure_smoke_test.dart',
+    );
+    final List<FilePath> screenshotTestFiles = <FilePath>[];
+    final List<FilePath> unitTestFiles = <FilePath>[];
+
+    for (final FilePath testFilePath in testFiles) {
+      if (!testFilePath.absolute.endsWith('_test.dart')) {
+        // Not a test file at all. Skip.
+        continue;
+      }
+      if (testFilePath == failureSmokeTestPath) {
+        // A smoke test that fails on purpose. Skip.
+        continue;
+      }
+
+      // All files under test/golden_tests are considered golden tests.
+      final bool isUnderGoldenTestsDirectory =
+          pathlib.split(testFilePath.relativeToWebUi).contains('golden_tests');
+      // Any file whose name ends with "_golden_test.dart" is run as a golden test.
+      final bool isGoldenTestFile = pathlib
+          .basename(testFilePath.relativeToWebUi)
+          .endsWith('_golden_test.dart');
+      if (isUnderGoldenTestsDirectory || isGoldenTestFile) {
+        screenshotTestFiles.add(testFilePath);
+      } else {
+        unitTestFiles.add(testFilePath);
+      }
+    }
+
+    // This test returns a non-zero exit code on purpose. Run it separately.
+    if (testFiles.contains(failureSmokeTestPath)) {
+      await _runTestBatch(
+        testFiles: <FilePath>[failureSmokeTestPath],
+        browserEnvironment: browserEnvironment,
+        concurrency: 1,
+        expectFailure: true,
+        isDebug: isDebug,
+        doUpdateScreenshotGoldens: doUpdateScreenshotGoldens,
+      );
+    }
+
+    // Run all unit-tests as a single batch and with high concurrency.
+    if (unitTestFiles.isNotEmpty) {
+      await _runTestBatch(
+        testFiles: unitTestFiles,
+        browserEnvironment: browserEnvironment,
+        concurrency: 10,
```

Comment:
Can we use a constant for this value or calculate it based on available CPUs?

Reply:
Done.